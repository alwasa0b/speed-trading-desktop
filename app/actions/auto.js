import {
  UPDATE_TIME_INTERVAL,
  UPDATE_UNDER_BID_PRICE,
  UPDATE_OVER_ASK_PRICE,
  UPDATE_QUANTITY_AUTO,
  NUMBER_OF_RUNS,
  UPDATE_NUMBER_OF_OPEN_ORDERS,
  UPDATE_PAUSE_PRICE,
  CLEAR
} from "../constants/auto";

import {
  START_WORKER,
  WORKER_STARTED,
  STOP_WORKER,
  WORKER_STOPPED,
  PAUSE_WORKER,
  UPDATE_WORKER,
  WORKER_UPDATED,
  WORKER_PAUSED,
  WORKER_RESUMED,
  RESUME_WORKER,
  PANIC
} from "../constants/messages";

const { ipcRenderer } = require("electron");

export const update_pause_price = payload => ({
  type: UPDATE_PAUSE_PRICE,
  payload
});

export const update_number_of_runs = payload => ({
  type: NUMBER_OF_RUNS,
  payload
});

export const update_number_of_open_orders = payload => ({
  type: UPDATE_NUMBER_OF_OPEN_ORDERS,
  payload
});

export const update_time_interval = payload => ({
  type: UPDATE_TIME_INTERVAL,
  payload
});

export const update_under_bid_price = payload => ({
  type: UPDATE_UNDER_BID_PRICE,
  payload
});

export const update_over_ask_price = payload => ({
  type: UPDATE_OVER_ASK_PRICE,
  payload
});

export const update_quantity = payload => ({
  type: UPDATE_QUANTITY_AUTO,
  payload
});

export const startStopWorker = () => (dispatch, getState) => {
  const { auto_order } = getState();
  if (auto_order.running) stop_worker(dispatch, getState);
  else start_worker(dispatch, getState);
};

const start_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  const { auto_order, messages } = getState();
  const { instrument, symbol, price, high } = messages.price;

  ipc.once(WORKER_STARTED, async event => {
    dispatch({ type: WORKER_STARTED });
  });

  ipc.once(WORKER_STOPPED, async event => {
    dispatch({ type: WORKER_STOPPED });
  });

  await ipc.send(START_WORKER, {
    ...auto_order,
    messages: [],
    instrument,
    symbol,
    price,
    high
  });
};

const stop_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  const { auto_order, messages } = getState();
  const { instrument, symbol } = messages.price;

  await ipc.send(STOP_WORKER, {
    ...auto_order,
    instrument,
    symbol
  });
};

export const pause_resume_worker = () => (dispatch, getState) => {
  const { auto_order } = getState();
  if (auto_order.paused) resume_worker(dispatch, getState);
  else pause_worker(dispatch, getState);
};

const resume_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  await ipc.send(RESUME_WORKER);
};

const pause_worker = async (dispatch, getState, ipc = ipcRenderer) => {
  ipc.once(WORKER_PAUSED, async event => {
    dispatch({ type: WORKER_PAUSED });
  });
  ipc.once(WORKER_RESUMED, async event => {
    dispatch({ type: WORKER_RESUMED });
  });
  await ipc.send(PAUSE_WORKER);
};

export const update_worker = () => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  const { auto_order, messages } = getState();
  const { instrument, symbol } = messages.price;

  ipc.once(WORKER_UPDATED, async event => {
    dispatch({ type: WORKER_UPDATED });
  });

  await ipc.send(UPDATE_WORKER, {
    ...auto_order,
    instrument,
    symbol
  });
};

export const panic = () => async (dispatch, getState, ipc = ipcRenderer) => {
  await ipc.send(PANIC);
};

export const clear = () => ({ type: CLEAR });
