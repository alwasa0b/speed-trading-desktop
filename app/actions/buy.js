import {
  PLACE_BUY_REQUEST_SUCCESS,
  UPDATE_QUANTITY,
  UPDATE_SELL_ORDER_TYPE,
  UPDATE_BUY_ORDER_TYPE,
  UPDATE_BUY_PRICE,
  UPDATE_SELL_PRICE,
  PLACE_BUY_REQUEST
} from "../constants/buy";
const { ipcRenderer } = require("electron");

const buy_order_success = () => ({ type: PLACE_BUY_REQUEST_SUCCESS });
export const place_buy_order = () => async (dispatch, getState) => {
  const { buy_order, messages } = getState();
  const { instrument, symbol } = messages.price;

  await ipcRenderer.send(PLACE_BUY_REQUEST, {
    ...buy_order,
    instrument,
    symbol
  });

  ipcRenderer.once(PLACE_BUY_REQUEST_SUCCESS, async (event, data) => {
    dispatch(buy_order_success(data));
  });
};

export const update_quantity = ({ quantity }) => ({
  type: UPDATE_QUANTITY,
  quantity
});

export const update_sell_order_type = ({ sell_order_type }) => ({
  type: UPDATE_SELL_ORDER_TYPE,
  sell_order_type
});

export const update_buy_order_type = ({ buy_order_type }) => ({
  type: UPDATE_BUY_ORDER_TYPE,
  buy_order_type
});

export const update_buy_price = ({ buy_price }) => ({
  type: UPDATE_BUY_PRICE,
  buy_price
});

export const update_sell_price = ({ sell_price }) => ({
  type: UPDATE_SELL_PRICE,
  sell_price
});
