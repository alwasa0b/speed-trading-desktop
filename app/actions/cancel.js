import {
  PLACE_CANCEL_REQUEST,
  PLACE_CANCEL_REQUEST_SUCCESS
} from "../constants/cancel";

import { ipcRenderer } from "electron";

const cancel_order_success = () => ({ type: PLACE_CANCEL_REQUEST_SUCCESS });
export const place_cancel_order = ({ cancel_order }) => async dispatch => {
  await ipcRenderer.send(PLACE_CANCEL_REQUEST, cancel_order);

  ipcRenderer.once(PLACE_CANCEL_REQUEST_SUCCESS, async (event, data) => {
    dispatch(cancel_order_success(data));
  });
};
