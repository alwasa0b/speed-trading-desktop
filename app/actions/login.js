import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  UPDATE_USERNAME,
  UPDATE_PASSWORD
} from "../constants/login";

import { ipcRenderer } from "electron";

const login_success = () => ({ type: LOGIN_SUCCESS });
const login_failed = () => ({ type: LOGIN_FAILURE });

export const login = () => async (dispatch, getState) => {
  await ipcRenderer.send(LOGIN_REQUEST, { login: getState().login });
  ipcRenderer.once(LOGIN_SUCCESS, async (event, data) => {
    dispatch(login_success());
  });
};

export const update_username = ({ username }) => ({
  type: UPDATE_USERNAME,
  username
});

export const update_password = ({ password }) => ({
  type: UPDATE_PASSWORD,
  password
});
