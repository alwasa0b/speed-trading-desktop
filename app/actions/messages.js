import {
  UPDATE_POSITIONS,
  UPDATE_ORDERS,
  UPDATE_PRICE,
  PRICE_UPDATED,
  POSITIONS_UPDATED,
  ORDERS_UPDATED
} from "../constants/messages";

const { ipcRenderer } = require("electron");

export const update_positions = () => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  await ipc.removeAllListeners(POSITIONS_UPDATED);
  await ipc.send(UPDATE_POSITIONS);

  ipc.on(POSITIONS_UPDATED, async (event, data) => {
    dispatch({ type: POSITIONS_UPDATED, data });
  });
};

export const update_orders = () => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  await ipc.removeAllListeners(ORDERS_UPDATED);
  await ipc.send(UPDATE_ORDERS);

  ipc.on(ORDERS_UPDATED, async (event, data) => {
    dispatch({ type: ORDERS_UPDATED, data });
  });
};

export const update_price = ({ symbol }) => async (
  dispatch,
  getState,
  ipc = ipcRenderer
) => {
  await ipc.removeAllListeners(PRICE_UPDATED);
  await ipc.send(UPDATE_PRICE, { symbol });

  ipc.on(PRICE_UPDATED, async (event, data) => {
    dispatch({ type: PRICE_UPDATED, data });
  });
};
