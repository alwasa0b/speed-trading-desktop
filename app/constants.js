export const userConstants = {
    LOGIN_REQUEST: "USERS_LOGIN_REQUEST",
    LOGIN_SUCCESS: "USERS_LOGIN_SUCCESS",
    LOGIN_FAILURE: "USERS_LOGIN_FAILURE",
    LOGOUT: "USERS_LOGOUT"
  };
  
  export const actions = {
    BUY_REQUEST: "BUY_REQUEST",
    SELL_REQUEST: "SELL_REQUEST",
    CANCEL_REQUEST: "CANCEL_REQUEST"
  };
  
  export const messages = {
    POSITIONS: "POSITIONS",
    ORDERS: "ORDERS",
    PRICE: "PRICE",
    UPDATE_PRICE: "SERVER/UPDATE_PRICE",
    UPDATE_POSITIONS: "SERVER/UPDATE_POSITIONS",
    UPDATE_ORDERS: "SERVER/UPDATE_ORDERS"
  };
  