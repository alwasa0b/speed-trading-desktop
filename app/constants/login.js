export const LOGIN_REQUEST = "LOGIN_REQUEST";
export const LOGIN_SUCCESS = "LOGIN_SUCCESS";
export const LOGIN_FAILURE = "LOGIN_FAILURE";
export const LOGOUT_REQUEST = "LOGOUT_REQUEST";
export const LOGOUT_SUCCESS = "LOGOUT_SUCCESS";
export const UPDATE_USERNAME = "UPDATE_USERNAME";
export const UPDATE_PASSWORD = "UPDATE_PASSWORD";
