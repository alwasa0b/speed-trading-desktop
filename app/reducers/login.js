import { UPDATE_USERNAME, UPDATE_PASSWORD } from "../constants/login";

export default (
  state = {
    username: "",
    password: ""
  },
  action
) => {
  switch (action.type) {
    case UPDATE_USERNAME:
      return { ...state, username: action.username };
    case UPDATE_PASSWORD:
      return { ...state, password: action.password };
    default:
      return state;
  }
}
