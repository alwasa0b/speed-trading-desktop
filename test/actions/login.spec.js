import sinon, { spy } from "sinon";
import * as actions from "../../app/actions/login";
import { LOGIN_REQUEST, LOGIN_SUCCESS } from "../../app/constants/login";

describe("login actions", () => {
  it("update_username should create update_username action", () => {
    expect(
      actions.update_username({
        username: "test_username"
      })
    ).toMatchSnapshot();
  });

  it("update_password should create update_password action", () => {
    expect(
      actions.update_password({
        password: "test_password"
      })
    ).toMatchSnapshot();
  });

  it("should call ipc.send with stop_order", () => {
    const fn = actions.login();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const ipc = { send: spy(), once: spy() };

    fn(
      dispatch,
      () => ({ login: { username: "test_user", password: "test_password" } }),
      ipc
    );

    expect(
      ipc.send.calledWith(LOGIN_REQUEST, {
        login: { username: "test_user", password: "test_password" }
      })
    ).toBe(true);
  });

  it("should start listening to ipc.once", done => {
    const fn = actions.login();
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    const ipc = { send: spy(), once: spy() };

    fn(
      dispatch,
      () => ({ login: { username: "test_user", password: "test_password" } }),
      ipc
    );

    setTimeout(() => {
      expect(
        ipc.once.calledWith(LOGIN_SUCCESS, sinon.match.typeOf("function"))
      ).toBe(true);
      done();
    }, 5);
  });
});
