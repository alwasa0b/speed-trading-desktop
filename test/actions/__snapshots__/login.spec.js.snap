// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login actions update_password should create update_password action 1`] = `
Object {
  "password": "test_password",
  "type": "UPDATE_PASSWORD",
}
`;

exports[`login actions update_username should create update_username action 1`] = `
Object {
  "type": "UPDATE_USERNAME",
  "username": "test_username",
}
`;
